src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperAttributeValueVisitor.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        /// <summary>
+        /// Writes code for the given <paramref name="chunk"/>.
+        /// </summary>
+        /// <param name="chunk">The <see cref="ResolveUrlChunk"/> to render.</param>
+        /// <remarks>
+        /// Allowed to support future C# extensions. Likely "~/..." will lead to a C# compilation error but that is up
+        /// to the compiler.
+        /// </remarks>
+        protected override void Visit(ResolveUrlChunk chunk)
+        {
+            RenderCode(chunk.Url, (Span)chunk.Association);
+        }
+
+        /// <summary>
+        /// Writes code for the given <paramref name="chunk"/>.
+        /// </summary>
+        /// <param name="chunk">The <see cref="SectionChunk"/> to render.</param>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown unconditionally to inform user of unexpected <c>@section</c> directive.
+        /// </exception>
+        protected override void Visit(SectionChunk chunk)
+        {
+            throw new InvalidOperationException(RazorResources.FormatParseError_Sections_Cannot_Be_Nested(
+                SyntaxConstants.TransitionCharacter + SyntaxConstants.CSharp.SectionKeyword));
+        }
+
+        /// <summary>
+        /// Writes code for the given <paramref name="chunk"/>.
+        /// </summary>
+        /// <param name="chunk">The <see cref="SetLayoutChunk"/> to render.</param>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown unconditionally to inform user of unexpected <c>@layout</c> directive.
+        /// </exception>
+        protected override void Visit(SetLayoutChunk chunk)
+        {
+            throw new InvalidOperationException(RazorResources.FormatParseError_Unexpected(
+                SyntaxConstants.TransitionCharacter + SyntaxConstants.CSharp.LayoutKeyword));
+        }
+
+        /// <summary>
+        /// Writes code for the given <paramref name="chunk"/>.
+        /// </summary>
+        /// <param name="chunk">The <see cref="StatementChunk"/> to render.</param>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown unconditionally to inform user of unexpected code block.
+        /// </exception>
+        protected override void Visit(StatementChunk chunk)
+        {
+            throw new InvalidOperationException(RazorResources.ParseError_Unexpected_Nested_CodeBlock);
+        }
+
+        /// <summary>
+        /// Writes code for the given <paramref name="chunk"/>.
+        /// </summary>
+        /// <param name="chunk">The <see cref="TemplateChunk"/> to render.</param>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown unconditionally to inform user of unexpected template e.g. <c>@&lt;p&gt;paragraph@&lt;/p&gt;</c>.
+        /// </exception>
+        protected override void Visit(TemplateChunk chunk)
+        {
+            throw new InvalidOperationException(RazorResources.ParseError_InlineMarkup_Blocks_Cannot_Be_Nested);
```

Comment:
Not that we expose. We'll need to flow the shared `ParserErrorSink` through, this will be the first bit of CodeGeneration that generates errors in "expected" cases.